pkg/reconciler/buildrun/resources/sources/bundle_core.go

```diff
@@ -0,0 +1,227 @@
+		return nil, err
+	}
+
+	image, err := mutate.AppendLayers(empty.Image, bundleLayer)
+	if err != nil {
+		return nil, err
+	}
+
+	return mutate.Time(image, time.Unix(0, 0))
+}
+
+func bundleDirectory(directory string) (v1.Layer, error) {
+	var split = func(path string) []string { return strings.Split(path, string(filepath.Separator)) }
+
+	var patterns []gitignore.Pattern
+	if file, err := os.Open(filepath.Join(directory, shpIgnoreFilename)); err == nil {
+		defer file.Close()
+
+		domain := split(directory)
+		scanner := bufio.NewScanner(file)
+		for scanner.Scan() {
+			line := strings.TrimSpace(scanner.Text())
+			if len(line) != 0 && !strings.HasPrefix(line, "#") {
+				patterns = append(patterns, gitignore.ParsePattern(line, domain))
+			}
+		}
+
+		if err := scanner.Err(); err != nil {
+			return nil, err
+		}
+	}
+
+	matcher := gitignore.NewMatcher(patterns)
+
+	var buf bytes.Buffer
+	var tw = tar.NewWriter(&buf)
+	defer tw.Close()
+
+	err := filepath.Walk(directory, func(path string, info os.FileInfo, err error) error {
+		// Bail out on path errors
+		if err != nil {
+			return err
+		}
+
+		if matcher.Match(split(path), info.IsDir()) {
+			if info.IsDir() {
+				return filepath.SkipDir
+			}
+
+			return nil
+		}
+
+		inBundlePath, err := tarPath(directory, path)
+		if err != nil {
+			return err
+		}
+
+		switch {
+		case info.Mode().IsDir():
+			return tw.WriteHeader(&tar.Header{
+				Name:     inBundlePath,
+				Typeflag: tar.TypeDir,
+				Mode:     int64(info.Mode()),
+			})
+
+		case info.Mode().IsRegular():
+			file, err := os.Open(path)
+			if err != nil {
+				return err
+			}
+
+			defer file.Close()
+
+			if err := tw.WriteHeader(&tar.Header{
+				Name:     inBundlePath,
+				Typeflag: tar.TypeReg,
+				Mode:     int64(info.Mode()),
+				Size:     info.Size()}); err != nil {
+				return err
+			}
+
+			_, err = io.Copy(tw, file)
+			return err
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return tarball.LayerFromReader(&buf)
+}
+
+func tarPath(baseDir string, path string) (string, error) {
+	rel, err := filepath.Rel(baseDir, path)
+	if err != nil {
+		return "", err
+	}
+
+	return filepath.Join(bundleBase, rel), nil
+}
+
+// Unbundle a container image layer content into a local directory. Analog to
+// the sources.Bundle function, optional remote.Option can be used to configure
+// settings for the image pull, i.e. access credentials.
+func Unbundle(ref name.Reference, targetPath string, options ...remote.Option) error {
+	desc, err := remote.Get(ref, options...)
+	if err != nil {
+		return err
+	}
+
+	image, err := desc.Image()
+	if err != nil {
+		return err
+	}
+
+	rc := mutate.Extract(image)
+	defer rc.Close()
+
+	var tr = tar.NewReader(rc)
+	for {
+		header, err := tr.Next()
+		switch {
+		case err == io.EOF:
+			return nil
+
+		case err != nil:
+			return err
+
+		case header == nil:
+			continue
+		}
+
+		var target = filepath.Join(targetPath, header.Name)
+		switch header.Typeflag {
```

Comment:
What about the other types? Can an image contain a symlink (not necessarily bundled by our own code)? Is imo at least worth a `default` case that returns an error.

Reply:
I played around with it, but eventually decided to follow the pragmatic approach for now: Any symlink is dereferenced (similar to `tar -H` I think). Also, I added the default case to catch unsupported types.